### PR TITLE
Initial NLopt support

### PR DESCRIPTION
**Pull request by jschueller**
For experimentation

### CODE REVIEW

**Comment by regislebrun** [comment]
+1
It looks very promising! More work is needed eg to document the current implementation of the NLopt class (no docstring) or to provide a richer interface to the collection of algorithms provided by NLopt (currently the use of the SLSQP algorithm is hard-coded).

**Comment by jschueller** [comment]
@regislebrun added doc & more algos.

**Comment by sofianehaddad** [comment]
+1

File: lib/src/Base/Optim/OptimizationSolverImplementation.cxx [hunk 1]
```diff
@@ -30,6 +30,7 @@ OptimizationSolverImplementation::OptimizationSolverImplementation()
   : PersistentObject()
   , startingPoint_(NumericalPoint(0))
   , maximumIterationsNumber_(ResourceMap::GetAsUnsignedInteger( "OptimizationSolverImplementation-DefaultMaximumIteration" ))
+  , maximumEvaluationNumber_(ResourceMap::GetAsUnsignedInteger("OptimizationSolverImplementation-DefaultMaximumEvaluationNumber"))
```
Comment (dbarbier) [inline lib/src/Base/Optim/OptimizationSolverImplementation.cxx]: We should be consistent: either maximumIterationsNumber_/maximumEvaluationsNumber_ or maximumIterationNumber_/maximumEvaluationNumber_, I am not sure which one is right, it seems that we prefer singular form.

**Comment by jschueller** [comment]
@dbarbier, ok, I'll deprecate *IterationsNumber methods.

**Comment by dbarbier** [comment]
It may be nice to have a
```
 static bool NLopt::IsAvailable() const
```
method.

**Comment by dbarbier** [comment]
+1, looks good.

File: lib/src/Base/Optim/OptimizationSolverImplementation.hxx [hunk 2]
```diff
@@ -120,7 +124,13 @@ protected:
 private:
   NumericalPoint startingPoint_;
   OptimizationProblem  problem_;
-  UnsignedInteger    maximumIterationsNumber_; /**< Number of outermost iterations (in case of nested iterations) */
+
+  /** Number of outermost iterations (in case of nested iterations) */
+  UnsignedInteger maximumIterationsNumber_;
+
+  /** Maximum function calls */
+  UnsignedInteger maximumEvaluationNumber_;
```
Comment (dbarbier) [inline lib/src/Base/Optim/OptimizationSolverImplementation.hxx]: You forgot to handle this new member in load/save/repr methods.

File: python/test/t_NLopt_std.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                    problem.setEqualityConstraint(ot.NumericalMathFunction(['x1', 'x2', 'x3', 'x4'], ['eq'], ['x4-2']))
+                try:
+                    algo.setProblem(problem)
+                    algo.setStartingPoint(startingPoint)
+                    print('algo=', algo)
+                    algo.run()
+                    result = algo.getResult()
+                    print('x^=', printNumericalPoint(result.getOptimalPoint(), 4))
+                except:
+                    print('-- Not supported: algo=', algo.__class__.__name__, 'inequality=', inequality, 'equality=', equality)
```
Comment (dbarbier) [inline python/test/t_NLopt_std.py]: Replace `algo.__class__.__name__` by `algo.getClassName()`?

Comment (jschueller) [inline python/test/t_NLopt_std.py]: Why ? it work with every Python object, imho getClassName shouldn't even be provided in Python.

**Comment by jschueller** [comment]
@dbarbier, about NLopt::IsAvailable, I agree for the principle, but maybe this could be addressed globally (and separately) to apply to each optional dependency xml, tbb, muparser, ...

File: python/test/t_NLopt_std.py (hunk 3 above)
Comment (dbarbier) [inline python/test/t_NLopt_std.py]: For consistency reasons, we already use `getClassName` in few python tests, but no `__class__.__name__`.

Comment (jschueller) [inline python/test/t_NLopt_std.py]: ok

**Comment by jschueller** [comment]
@dbarbier I fixed the use of the new optimsolver attribute, and deprecated  [sg]etMaximumIterationsNumber

**Comment by regislebrun** [comment]
+1

**Comment by sofianehaddad** [comment]
I think you forgot to update ChangeLog

**Comment by jschueller** [comment]
@sofianehaddad fixed

**Pull request merged**